pkg/catalog/configmeta.go

```diff
@@ -177,20 +177,29 @@ func servicesForRole(
 	return result
 }
 
-// genconfigConnections will look at the cluster spec
+// genConfigConnections will look at the cluster spec
 // and generates a map of configmap type and corresponding
 // configmaps to be connected to the given cluster
-func genconfigConnections(
+func genConfigConnections(
 	cr *kdv1.KubeDirectorCluster,
-) (map[string]map[string]map[string]string, error) {
+) (map[string][]map[string]map[string]string, error) {
 
-	cmMap := make(map[string]map[string]string)
-	kdcm := make(map[string]map[string]map[string]string)
+	// Many connected configmaps can be of a given type, hence
+	// create a map of cmType:list of configmaps, where
+	// every configmap is a map of string and string
+	kdcm := make(map[string][]map[string]map[string]string)
 	for _, connectedCmName := range cr.Spec.Connections.ConfigMaps {
 		cm, err := observer.GetConfigMap(cr.Namespace, connectedCmName)
 		if kdConfigMapType, ok := cm.Labels[configMapType]; ok {
+			cmMap := make(map[string]map[string]string)
 			cmMap[connectedCmName] = cm.Data
```

Comment:
So the map at this level will always only have one key? Hmm.

I mean this is fine if this is the structure that works best for the consumer, but it seems like semantically the slice elements are a pair of name-and-map, rather than a map of maps.

Reply:
Ya, that is right, we look for a specific key for config map and secrets. We can add more key(s) in the future. Having said that this is already quite flexible. We can store all maps, secrets of a given type as a list, and it's intuitive for apps to query in those ways.

Reply:
We might not be talking about the same thing so let me get into some details...

1. We have a top-level map where key is "type", and value is a list.
2. In the list, each element is a map.
3. In this map, there is always only one entry. The key is the name of the connected resource, and the value is its data (which is another map).

It's level 3 that seems a little odd. A map that always only has one entry is really just a pair, which might more naturally represented as a list-of-2-things or a struct-with-2-fields.

Are you saying that that map at level 3 could have different keys in the future?

Reply:
ok, yes you are spot on. The 3rd level key is the name of the resource. We could put more stuff there like labels, annotations since only data is being propagated now. Does it make sense?

Reply:
Got it, thanks!

Reply:
Although what would the keys look like then? Currently the "lone key" could be anything. The resource itself could even be named "labels" in which case the key would be "labels".

If we did add other things to this map like labels and annotations what would the map then look like?

Reply:
Currently, the DS looks like
```
"security" // cmType : [  {"ldapinfo" :  
                                           {<k1 : v1>, <k2 : v2>}},
                                          {"kerberos" : 
                                            {<k1 : v1>, <k2 : v2>}}
                                     ] 
```
When we propagate labels and keys it should be something like -

```
"security" // cmType : [  {"ldapinfo/data" :  
                                           {<k1 : v1>, <k2 : v2>},
                                        {"ldapinfo/labels" : 
                                           {<k1 : v1>, <k2 : v2>},
                                        {"ldapinfo/annotations" : 
                                           {<k1 : v1>, <k2 : v2>}, 
                                       // and more metadata about the resource

                                     ] 
```

Reply:
Do you mean this:
```
"security" // cmType : [{"ldapinfo/data" : {<k1 : v1>, <k2 : v2>},
                         "ldapinfo/labels" : {<k1 : v1>, <k2 : v2>},
                         "ldapinfo/annotations" : {<k1 : v1>, <k2 : v2>},
                         // etc.
                        },

                        {"kerberos/data" : {<k1 : v1>, <k2 : v2>},
                         // etc.
                        }] 
```
I.e. the ldap/\* stuff are all keys in one of the 3rd-level maps.

I'm not 100% clear I guess on how the consumer is supposed to use this. Do they typically need to know the resource name? Will they be walking the keys and splitting each key on slash to figure out what the value is representing?

Maybe it should be something like
```
"security" // cmType : [{"metadata" : {<k1 : v1>, <k2 : v2>},
                         "data" : {<k1 : v1>, <k2 : v2>},
                         "labels" : {<k1 : v1>, <k2 : v2>},
                         "annotations" : {<k1 : v1>, <k2 : v2>}}]
```
where metadata has at least the resource name and could also carry kind and namespace for the heck of it. For this release we could populate the metadata and data maps.

But you know more than I do about how this is currently being used.

In any case I think we should establish the scheme now rather than planning to immediately change it in the next release, even if we're not yet putting labels or annotations in there.

Reply:
okay, yea, I overengineered in my example but you got the idea. Will adhere to your example structure